db/python/tables/sequence.py

```diff
@@ -279,21 +279,44 @@ async def get_sequence_type_numbers_for_project(self, project: ProjectId):
         """
 
         _query = """
-SELECT type, COUNT(*) as n
-FROM (
-    SELECT sq.type
-    FROM sample_sequencing sq
-    INNER JOIN sample s ON s.id = sq.sample_id
-    WHERE s.project = :project
-    GROUP BY s.id, sq.type
-) as s
-GROUP BY type
+            SELECT type, COUNT(*) as n
+            FROM (
+                SELECT sq.type
+                FROM sample_sequencing sq
+                INNER JOIN sample s ON s.id = sq.sample_id
+                WHERE s.project = :project
+                GROUP BY s.id, sq.type
+            ) as s
+            GROUP BY type
         """
 
         rows = await self.connection.fetch_all(_query, {'project': project})
 
         return {r['type']: r['n'] for r in rows}
 
+    async def get_sequence_type_numbers_by_batch_for_project(self, project: ProjectId):
+        """
+        This groups by batch.
+        """
+
+        _query = """
+        SELECT batch, type, count(*) AS n
+        FROM (
+            SELECT IFNULL(JSON_EXTRACT(sq.meta, '$.batch'), 'null') as batch, sq.type
+            FROM sample_sequencing sq
+            INNER JOIN sample s ON s.id = sq.sample_id
+            WHERE s.project = :project
+        ) as p
+        GROUP BY batch, type
+            """
+        rows = await self.connection.fetch_all(_query, {'project': project})
+        batch_result: Dict[str, Dict[str, str]] = {}
+        for batch, seqType, count in rows:
+            batch = batch.strip('\"')
+            batch_result.setdefault(seqType, {}).update(dict({batch: str(count)}))
```

Comment:
Consider making batch_result a defaultdictionary:

```suggestion
        batch_result: Dict[str, Dict[str, str]] = defaultdict(dict)
        for batch, seqType, count in rows:
            batch = batch if batch else "no-batch"
            batch = batch.strip('\"')
            batch_result[seqType][batch] = str(count)
```

Reply:
Fixed